### DISCUSSION TITLE
DB side LUA scripting

### DISCUSSION THREAD

**Discussion by aarthikrao** [aarthikrao/timeMachine#3] [Ideas]
Time machine can have DB side scripting using LUA. We can use this to write business logic and run transactions atomically at the database end. 

**Limitations**
- keys needed in the script will have to be confined to a single partition
- timeout to discourage long running queries.

**Comment by devansh42** [comment]
Hi, Exposing an LUA based interface to write business logic can expose timeMachine to abusive situations. Instead of this, in future we can expose a plugin API  just like Redis Plugins. That would be much safer, extensible and will help to ensure SLAs around timeMachine.

**Comment by devansh42** [comment]
With a proper plugin oriented architecture we can enable features like Alerts, Alarms, Rate Limiting, Logging, Tracing, etc. 
Alerts are my favourite. 😄

**Reply by aarthikrao** [reply]
We can have Lua scripting with limited scope. Just like redis Lua scripting. 

And yes. Plugins are a good idea. We should try to build them with the same redis ecosystem so that we can reuse them.

😸

**Comment by aarthikrao** [comment]
https://redis.io/docs/manual/programmability/

Redis runs the Lua script in a sandbox. This way it prevents misuse. 
Similarly, we can allow only certain operations.
